src/StreamJsonRpc/JsonMessageFormatter.cs

```diff
@@ -583,17 +583,55 @@ internal JsonRpcRequest(JsonMessageFormatter formatter)
                 {
-                    var obj = new JObject();
-                    foreach (KeyValuePair<string, object> property in this.NamedArguments)
+                    // Special support for accepting a single JToken instead of all parameters individually.
+                    if (parameters[0].ParameterType == typeof(JToken))
                     {
-                        obj.Add(new JProperty(property.Key, property.Value));
+                        var obj = new JObject();
+                        foreach (KeyValuePair<string, object> property in this.NamedArguments)
+                        {
+                            obj.Add(new JProperty(property.Key, property.Value));
+                        }
+
+                        typedArguments[0] = obj;
+                        return ArgumentMatchResult.Success;
                     }
 
-                    typedArguments[0] = obj;
-                    return ArgumentMatchResult.Success;
+                    // Check if it is constructable with default constructor
+                    if (parameters[0].ParameterType.GetConstructor(Type.EmptyTypes) != null && !parameters[0].ParameterType.IsAbstract)
+                    {
+                        // If the method is expecting only one parameter we should try to create the expected object with the given arguments
+                        var obj = Activator.CreateInstance(parameters[0].ParameterType);
+                        bool match = false;
+
+                        // Loop value-key pairs from arguments
+                        foreach (KeyValuePair<string, object> property in this.NamedArguments)
+                        {
+                            // If argument name matches property name, assign the value to that property
+                            PropertyInfo propertyInfo = parameters[0].ParameterType.GetProperty(property.Key);
+                            JToken value = (JToken)property.Value;
+                            if (propertyInfo != null)
+                            {
+                                propertyInfo.SetValue(obj, value.ToObject(propertyInfo.PropertyType, this.formatter.JsonSerializer));
+                                match = true;
+                            }
+                            else
+                            {
+                                FieldInfo fieldInfo = parameters[0].ParameterType.GetField(property.Key);
+                                if (fieldInfo != null)
+                                {
+                                    fieldInfo.SetValue(obj, value.ToObject(fieldInfo.FieldType, this.formatter.JsonSerializer));
+                                    match = true;
+                                }
+                            }
+                        }
+
+                        typedArguments[0] = obj;
+
+                        // If one of the arguments matched the parameters we return ArgumentMatchResult.Success
```

Comment:
This criteria of allowing just one parameter/property name match sounds like you're mixing in #288 with #289. So if nothing else I think your PR description should call out that it closes #288 as well. But IMO this should be its own opt-in. Perhaps another `JsonRpcMethodAttribute` property that you add to `AllowFuzzyMatching`.

Reply:
I see, so for #289 should it be something similar to what we do for JToken, only difference would be that we deserialize it to the specific type before assigning to typedArguments[0] right? Something like:
`typedArguments[0] = obj.ToObject(parameters[0].ParameterType);`

---
In reply to: [336723042](https://github.com/microsoft/vs-streamjsonrpc/pull/347#discussion_r336723042) [](ancestors = 336723042)

Reply:
Ya, that sounds about right.
One trick may be though is determining when to allow this. We don't currently tunnel properties on `JsonRpcMethodAttribute` to the formatter, which may be required if you go with the way I was proposing.

Reply:
I refactored the logic on the latest update